### PR TITLE
[DROOLS-5730] improve performances of arguments fetch in executable m…

### PR DESCRIPTION
**Pull request by mariofusco**

### CODE REVIEW

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/KiePackagesBuilder.java [hunk 1]
```diff
@@ -375,30 +375,50 @@ private void addQueryPattern( Query query, QueryImpl queryImpl, RuleContext ctx
...
+        String[] requiredDeclarationNames = new String[consequenceVars.length];
+        Declaration[] requiredDeclarations = ruleHasFirstLevelOr ? null : new Declaration[consequenceVars.length];
+        for (int i = 0; i < consequenceVars.length; i++) {
+            requiredDeclarationNames[i] = consequenceVars[i].getName();
+            if (!ruleHasFirstLevelOr) {
+                requiredDeclarations[i] = ctx.getRule().getDeclaration( requiredDeclarationNames[i] );
+            }
+        }
+
+        ctx.getRule().setRequiredDeclarationsForConsequence( name, requiredDeclarationNames );
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/KiePackagesBuilder.java]: Is it possible to wrap this logic in a method to improve readability? 
```java
/**
 * When rule has only AND as top level, it is possible to precalculate requiredDeclarations as optimization 
 */
private Declaration[] getRequiredDeclarationsIfPossible( RuleContext ctx, Consequence consequence, String name ) {
        boolean ruleHasFirstLevelOr = ruleHasFirstLevelOr(ctx.getRule());
        Variable[] consequenceVars = consequence.getDeclarations();
        String[] requiredDeclarationNames = new String[consequenceVars.length];
        Declaration[] requiredDeclarations = ruleHasFirstLevelOr ? null : new Declaration[consequenceVars.length];
        for (int i = 0; i < consequenceVars.length; i++) {
            requiredDeclarationNames[i] = consequenceVars[i].getName();
            if (!ruleHasFirstLevelOr) {
                requiredDeclarations[i] = ctx.getRule().getDeclaration( requiredDeclarationNames[i] );
            }
        }

        ctx.getRule().setRequiredDeclarationsForConsequence( name, requiredDeclarationNames );
        return requiredDeclarations;
}

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java [hunk 2]
```diff
@@ -43,23 +52,28 @@ public String getName() {
 
     @Override
     public void evaluate( KnowledgeHelper knowledgeHelper, WorkingMemory workingMemory ) throws Exception {
-        Declaration[] declarations = ((RuleTerminalNode)knowledgeHelper.getMatch().getTuple().getTupleSink()).getRequiredDeclarations();
-        Object[] facts = declarationsToFacts( knowledgeHelper, workingMemory, knowledgeHelper.getTuple(), declarations, consequence.getVariables(), consequence.isUsingDrools() );
+        Object[] facts;
+        if ( this.declarations == null ) {
+            Declaration[] declarations = (( RuleTerminalNode ) knowledgeHelper.getMatch().getTuple().getTupleSink()).getRequiredDeclarations();
+            facts = declarationsToFacts( knowledgeHelper, ( InternalWorkingMemory ) workingMemory, knowledgeHelper.getTuple(), declarations, consequence.getVariables(), consequence.isUsingDrools() );
+        } else {
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java:62]:
```suggestion
        // declarations is not null when first level rule is AND so it is possible to calculate them upfront
        } else {
```

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java [hunk 3]
```diff
@@ -69,19 +83,184 @@ public void evaluate( KnowledgeHelper knowledgeHelper, WorkingMemory workingMemo
...
         }
         return facts;
     }
 
-    private static InternalFactHandle getOriginalFactHandle(InternalFactHandle handle) {
-        InternalFactHandle linkedFH = handle.isEvent() ? ((EventFactHandle )handle).getLinkedFactHandle() : null;
+    private static InternalFactHandle getOriginalFactHandle( InternalFactHandle handle ) {
+        if ( !handle.isEvent() ) {
+            return handle;
+        }
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java:104]: Is this an additional/unrelated bugfix/improvement?

Comment (mariofusco) [reply drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java:104]: If look at the old code
```
    private static InternalFactHandle getOriginalFactHandle( InternalFactHandle handle ) {
        InternalFactHandle linkedFH = handle.isEvent() ? ((EventFactHandle )handle).getLinkedFactHandle() : null;
        return linkedFH != null ? linkedFH : handle;
    }

```
and at the new one
```
    private static InternalFactHandle getOriginalFactHandle( InternalFactHandle handle ) {
        if ( !handle.isEvent() ) {
            return handle;
        }
        InternalFactHandle linkedFH = (( EventFactHandle ) handle).getLinkedFactHandle();
        return linkedFH != null ? linkedFH : handle;
    }
```
you will realize that the 2 codes are functionally identical: if the handle is not an event nothing has to be done and the handle can be used as it is. I just wanted to make this thing more explicit and avoid a useless second null check in that case.

**Comment by mariofusco** [comment]
jenkins execute cdb

**Review by lucamolteni — APPROVED** [review]

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/KiePackagesBuilder.java [hunk 4]
```diff
@@ -375,30 +375,54 @@ private void addQueryPattern( Query query, QueryImpl queryImpl, RuleContext ctx
...
+    private void processConsequences( RuleContext ctx, Map<String, Consequence> consequences ) {
+        for (Map.Entry<String, Consequence> entry : consequences.entrySet()) {
             processConsequence( ctx, entry.getValue(), entry.getKey() );
         }
     }
 
     private void processConsequence( RuleContext ctx, Consequence consequence, String name ) {
+        // If there's an OR in the rule the fired tuple hasn't fixed structure and size because it dependens
+        // on which branch of the OR gets activated. In this case no optimization is possible and it's usless
+        // to precalculate the declartions at compile time.
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/KiePackagesBuilder.java]: Can you please first introduce the optimization and then explain why it is not applicable to `OR`? :)

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java [hunk 5]
```diff
@@ -69,19 +84,196 @@ public void evaluate( KnowledgeHelper knowledgeHelper, WorkingMemory workingMemo
...
+            if ( var.isFact() ) {
+                factSuppliers.add( new TupleFactSupplier(factsOffset, declarations[declrCounter++], consequence.isUsingDrools()) );
+            } else {
+                facts[factsOffset] = workingMemory.getGlobal( var.getName() );
+                factSuppliers.add( new GlobalSupplier(factsOffset, var.getName()) );
+            }
+            factsOffset++;
+        }
+
+        // At this point the FactSuppliers (each of them supplying a single argument to be passed to the consequence)
```
Comment (danielezonca) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/consequence/LambdaConsequence.java:147]: Can you refer the optimization here too (in the future we will not have the PR to see the whole code together :) )?

**Comment by mariofusco** [comment]
jenkins execute cdb

**Review by danielezonca — APPROVED** [review]

**Pull request merged**